worker/caasfirewaller/worker.go

```diff
@@ -97,41 +102,63 @@ func (p *firewaller) loop() error {
 			if !ok {
 				return errors.New("watcher closed channel")
 			}
-			for _, appID := range apps {
-				appLife, err := p.config.LifeGetter.Life(appID)
+			for _, appName := range apps {
+				// If charm is (now) a v2 charm, skip processing.
+				format, err := p.charmFormat(appName)
+				if errors.IsNotFound(err) {
+					p.config.Logger.Debugf("application %q removed", appName)
+					continue
+				} else if err != nil {
+					return errors.Trace(err)
+				}
+				if format >= corecharm.FormatV2 {
```

Comment:
This logic is run when the worker first starts up and is notified of existing application, or when a new application is added. Thus the "(now)" in the comment above is sort of irrelevant. What's missing is a watcher to detect the change of an application from v1 to v2 while juju is running. And when we detect that, it's not sufficient to just skip processing - we also need to stop any existing v1 worker for that app.

Reply:
Oh, I just noticed the application change watcher in the app worker so the above comment is incorrect; this look like they are good.